### DISCUSSION TITLE
Problem to connect RMq with Address resolver

### DISCUSSION THREAD

**Discussion by MikhailTushev** [rabbitmq/rabbitmq-stream-dotnet-client#302] [Q&A] [answered]
I have 3 hosts (nodes of rmq), like this: host1.com , host2.com , host3.com . 
How can I connect to rabbitmq using an address resolver (or something else)?
I tried first host write to the address resolver from the 3rd host. When the second endpoint is leader (host2:5552), the consumer/producer is not connected because it compares the resolver addresses - host1:5552 with leader - host2:5552.
What can i do?

**Comment by Gsantomaggio** [comment]
The client needs to resolve the hostnames. If the client can resolve the host names you use the endpoint list:

```
  var config = new StreamSystemConfig
            {
                UserName = "test",
                Password = "test",
                Endpoints = new List<EndPoint>()
                {
                    new DNSEndpoit(host1),
                    new DNSEndpoit(host2),
                    new DNSEndpoithost3),

                }
            };
 ````


if there is a load balancer, you need the  address resolver  like:

```
  var loadBalancer = new AddressResolver(new IPEndPoint(IPAddress.Parse("111.1111.111.111"), 5552));
            var config = new StreamSystemConfig
            {
                UserName = "test",
                Password = "test",
                AddressResolver = loadBalancer,
                Endpoints = new List<EndPoint>()
                {
                                    loadBalancer.EndPoint
                }
            };
```

**Reply by MikhailTushev** [reply]
but in the first example, for rmq needs setting advertised_host for all of nodes? 
because without it i have exception "Unknown host" in method GetEndPoint in LookupConnection:
` var hostEntry = await Dns.GetHostEntryAsync(broker.Host).ConfigureAwait(false);`

for example i try connect to localhost with opened port 5552 and enabled plugin. but without advertised_host not work

**Reply by Gsantomaggio** [reply, marked as answer]
You need to set the `advertised_host` in situations like:
https://rabbitmq.com/stream.html#advertised-host-port
( here for more details https://blog.rabbitmq.com/posts/2021/07/connecting-to-streams/ )

> By default, nodes return their hostname and listener port, which may be fine for most situations, but not always (proxy sitting between the cluster nodes and the clients, cluster nodes and/or clients running in containers, etc).


>"Unknown host" 

The client cannot look up the hostname, so you need `advertised_host` in that case. 

An example is using docker:
```
docker run -it --rm --name rabbitmq-stream-client-test \
		-p 5552:5552 -p 5672:5672 -p 15672:15672 \
		-e RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS="-rabbitmq_stream advertised_host localhost" \
		.....
```
Where the client is not able to lookup the docker hostname and use `localhost`

**Reply by MikhailTushev** [reply]
thank you<3!
